### PR TITLE
feat: [PL-25126]: service layer integration of create branch method

### PR DESCRIPTION
**Pull request by bhavya181**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33098)

### CODE REVIEW

**Comment by mohitg0795** [comment]
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/scm/GitRefType.java` line 26 at r1](https://reviewable.io/reviews/harness/harness-core/33098#-N2At8rG96Ui2JhgGzY5:-N2At8rHDigfouZphJ1F:b-voyykj) ([raw file](https://github.com/harness/harness-core/blob/2dccdcb59cfe21ab4826cbc0783b5e2058193f77/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/scm/GitRefType.java#L26)):*
> ```Java
>   BRANCH_LIST_WITH_DEFAULT_BRANCH,
>   REPOSITORY_DETAILS,
>   CREATE_NEW_BRANCH
> ```

Lets make it CREATE_BRANCH_V2, otherwise it becomes confusing whats "CREATE_BRANCH" and "CREATE_NEW_BRANCH" which are essentially the same.

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098)* status: 14 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/scm/GitRefType.java` line 26 at r1](https://reviewable.io/reviews/harness/harness-core/33098#-N2At8rG96Ui2JhgGzY5:-N2BX7mI2difNU1nKCmi:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2dccdcb59cfe21ab4826cbc0783b5e2058193f77/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/scm/GitRefType.java#L26)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Lets make it CREATE_BRANCH_V2, otherwise it becomes confusing whats "CREATE_BRANCH" and "CREATE_NEW_BRANCH" which are essentially the same.
</blockquote></details>

Done.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> Reviewed 7 of 13 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098#-:-N2BXvNi5MnrEhDwdGTi:bnfp4nl)* status: 14 file(s) to review. 1 of 2 approvals obtained.

**Review by DeepakPatankar — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098)* status: 18 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java` line 699 at r3](https://reviewable.io/reviews/harness/harness-core/33098#-N2Bicps6cPEYKTKJ2hs:-N2Bicps6cPEYKTKJ2ht:b-e99igy) ([raw file](https://github.com/harness/harness-core/blob/4162d57c769f0d8ad15bf32b332d75b7785ad038/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java#L699)):*
> ```Java
>       return CreateBranchResponse.parseFrom(scmGitRefTaskResponseData.getCreateBranchResponse());
>     } catch (InvalidProtocolBufferException e) {
>       String errorMsg = String.format(errorMessage, "create branch", scope.getAccountIdentifier(),
> ```

rename errorMessage to errorFormat

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java [hunk 1]
```diff
@@ -8,29 +8,51 @@
...
 @OwnedBy(PL)
 public class BitbucketCreateBranchScmApiErrorHandler implements ScmApiErrorHandler {
+  public static final String CREATE_BRANCH_FAILED_MESSAGE = "The requested branch could not be created on Bitbucket. ";
+  public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_HINT =
+      "Please check the requested branch name if it is valid and does not exists.";
+  public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_EXPLANATION = "Possible reasons can be:\n"
+      + "1. A branch with given name already exists in the remote Bitbucket repository.\n"
+      + "2. The given branch name is invalid.";
+  public static final String CREATE_BRANCH_NOT_FOUND_ERROR_HINT =
+      "Please check the requested Bitbucket repository / base branch, if they exist or not.";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java:33]: Please check if the requested Bitbucket repository or base branch exists.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java [hunk 2]
```diff
@@ -8,29 +8,51 @@
...
+  public static final String CREATE_BRANCH_NOT_FOUND_ERROR_EXPLANATION = "Possible reasons can be:\n"
+      + "1. The given bitbucket repository is invalid.\n"
+      + "2. The given base branch does not exists.";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java:36]: These are possible reasons for?

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java [hunk 3]
```diff
@@ -16,6 +16,6 @@
 @UtilityClass
 @OwnedBy(PL)
 public class ScmErrorHints {
-  public static final String INVALID_CREDENTIALS = "Please check your bitbucket credentials.";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java:19]: Please change the casing to Bitbucket across all the error messages.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java [hunk 4]
```diff
@@ -16,6 +16,6 @@
...
-  public static final String REPO_NOT_FOUND = "Please check your bitbucket repository.";
+  public static final String INVALID_CREDENTIALS = "Please check your Bitbucket credentials if they are valid or not.";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java:19]: Please check if your Bitbucket credentials are valid.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java [hunk 5]
```diff
@@ -16,6 +16,6 @@
...
+  public static final String REPO_NOT_FOUND = "Please check if requested Bitbucket repository exists or not.";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java:20]: Please check if the requested Bitbucket repository exists.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java [hunk 6]
```diff
@@ -8,34 +8,54 @@
...
+import io.harness.exception.ScmUnprocessableEntityException;
 import io.harness.exception.WingsException;
 import io.harness.gitsync.common.scmerrorhandling.handlers.ScmApiErrorHandler;
 
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
 @OwnedBy(PL)
 public class GithubCreateBranchScmApiErrorHandler implements ScmApiErrorHandler {
-  public static final String CREATE_BRANCH_FAILED = "The requested branch could not be created on Github. ";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java:24]: Let's provide a reason for such error messages.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java [hunk 7]
```diff
@@ -8,34 +8,54 @@
...
+  public static final String CREATE_BRANCH_FAILED_MESSAGE = "The requested branch could not be created on Github. ";
+  public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_HINT =
+      "Please check the requested base branch if it exists and new branch name, if it is valid and does not exists.";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java:28]: Refer to similar review comment above.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java [hunk 8]
```diff
@@ -8,34 +8,54 @@
...
+  public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_EXPLANATION = "Possible reasons can be:\n"
+      + "1. A branch with given name already exists in the remote Github repository.\n"
+      + "2. The given branch name is invalid.\n"
+      + "3. The given base branch does not exists.";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java:32]: Refer to previous comments.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java [hunk 9]
```diff
@@ -8,34 +8,54 @@
...
+  public static final String CREATE_BRANCH_CONFLICT_ERROR_HINT =
+      "Please check the requested Github base branch, if it exists or not.";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java:34]: Refer to previous comments.

File: 136-git-sync-manager/src/test/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImplTest.java [hunk 10]
```diff
@@ -240,4 +242,21 @@ public void testGetRepoDetails() {
     assertThat(getUserRepoResponse.getRepo().getName()).isEqualTo(repoName);
     assertThat(getUserRepoResponse.getRepo().getBranch()).isEqualTo(defaultBranch);
   }
+
+  @Test
+  @Owner(developers = BHAVYA)
+  @Category(UnitTests.class)
+  public void testCreateNewBranch() {
+    String errorMessage = "Repo not exist";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/test/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImplTest.java:250]: Repo does not exist.

File: 136-git-sync-manager/src/test/java/io/harness/gitsync/common/impl/ScmFacilitatorServiceImplTest.java [hunk 11]
```diff
@@ -125,4 +129,22 @@ public void testGetDefaultBranch() {
         accountIdentifier, orgIdentifier, projectIdentifier, connectorRef, repoName);
     assertThat(branchName).isEqualTo(defaultBranch);
   }
+
+  @Test
+  @Owner(developers = BHAVYA)
+  @Category(UnitTests.class)
+  public void testCreateNewBranch() {
+    String errorMessage = "Repo not exist";
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/test/java/io/harness/gitsync/common/impl/ScmFacilitatorServiceImplTest.java:137]: Refer to previous comments.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java [hunk 12]
```diff
@@ -8,29 +8,51 @@
...
+
+@Slf4j
 @OwnedBy(PL)
 public class BitbucketCreateBranchScmApiErrorHandler implements ScmApiErrorHandler {
+  public static final String CREATE_BRANCH_FAILED_MESSAGE = "The requested branch could not be created on Bitbucket. ";
+  public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_HINT =
+      "Please check the requested branch name if it is valid and does not exists.";
+  public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_EXPLANATION = "Possible reasons can be:\n"
+      + "1. A branch with given name already exists in the remote Bitbucket repository.\n"
+      + "2. The given branch name is invalid.";
```
[the remark this answers is not in the document]
Comment (bhavya181) [reply 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java:31]: Below is the whole error message shown to user, you can check that this string is explanation and hint is there above you asked for.

Please check that the requested Bitbucket branch name is valid and does not already exist.
The requested branch could not be created on Bitbucket. Possible reasons can be:
1. A branch with given name already exists in the remote Bitbucket repository.
2. The given branch name is invalid.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java (hunk 5 above)
Comment (bhavya181) [reply 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java]: done

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java (hunk 3 above)
Comment (bhavya181) [reply 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/ScmErrorHints.java:19]: this is deleted

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java [hunk 13]
```diff
@@ -8,29 +8,51 @@
...
 import io.harness.gitsync.common.scmerrorhandling.handlers.ScmApiErrorHandler;
 
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
 @OwnedBy(PL)
 public class BitbucketCreateBranchScmApiErrorHandler implements ScmApiErrorHandler {
+  public static final String CREATE_BRANCH_FAILED_MESSAGE = "The requested branch could not be created on Bitbucket. ";
+  public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_HINT =
+      "Please check the requested branch name if it is valid and does not exists.";
```
[the remark this answers is not in the document]
Comment (bhavya181) [reply 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java]: done

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java [hunk 14]
```diff
@@ -8,29 +8,51 @@
...
+import io.harness.exception.ScmUnexpectedException;
 import io.harness.exception.WingsException;
 import io.harness.gitsync.common.scmerrorhandling.handlers.ScmApiErrorHandler;
 
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
 @OwnedBy(PL)
 public class BitbucketCreateBranchScmApiErrorHandler implements ScmApiErrorHandler {
+  public static final String CREATE_BRANCH_FAILED_MESSAGE = "The requested branch could not be created on Bitbucket. ";
```
[the remark this answers is not in the document]
Comment (bhavya181) [reply 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucket/BitbucketCreateBranchScmApiErrorHandler.java:26]: so, this  a single string, which is combined with explanation, that is provoided below

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java (hunk 6 above)
Comment (bhavya181) [reply 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java:24]: explained above

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java (hunk 7 above)
Comment (bhavya181) [reply 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java]: done

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java (hunk 9 above)
Comment (bhavya181) [reply 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java]: done

File: 136-git-sync-manager/src/test/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImplTest.java (hunk 10 above)
Comment (bhavya181) [reply 136-git-sync-manager/src/test/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImplTest.java:250]: this is a test, so does not matter, it's for internal use only

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098)* status: 19 file(s) to review. 1 change requests. 13. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java` line 699 at r3](https://reviewable.io/reviews/harness/harness-core/33098#-N2Bicps6cPEYKTKJ2hs:-N2C7eLi5UYJ1bGfgMgf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/4162d57c769f0d8ad15bf32b332d75b7785ad038/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmDelegateFacilitatorServiceImpl.java#L699)):*
<details><summary><i>Previously, DeepakPatankar (Deepak Patankar) wrote…</i></summary><blockquote>

rename errorMessage to errorFormat
</blockquote></details>

Done.

**Review by rashminandasahoo — DISMISSED** [review]
:lgtm:

> Reviewed 5 of 13 files at r1, 4 of 7 files at r2, 3 of 6 files at r3, 6 of 6 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098#-:-N2CD6zS0lZxsRCJewAi:bnfp4nl)* status: 10 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> Reviewed 3 of 6 files at r3, 6 of 6 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098#-:-N2G0OgmDLegsaLsi9--:bnfp4nl)* status: 2 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by DeepakPatankar — COMMENTED** [review]
> Reviewed 5 of 13 files at r1, 4 of 7 files at r2, 3 of 6 files at r3, 6 of 6 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098)* status: 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by DeepakPatankar — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098)* status: :shipit: complete! ✓

**Review by abhinav-harness — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098)* status: :shipit: complete! ✓

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33098#-:-N2G3zTS7Dx8BOpe3bIj:bnfp4nl)* status: :shipit: complete! ✓
